### PR TITLE
SPARK-10329 Cost RDD in k-means|| initialization is not storage-efficient

### PR DESCRIPTION
**Pull request by hujy**
The PR is try to reduce memory cost and computation complexity of kmeans.

### CODE REVIEW

**Comment by HuJiayin** [comment]
cc @mengxr

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 1]
```diff
@@ -246,10 +248,23 @@ class KMeans private (
         if (initializationMode == KMeans.RANDOM) {
           initRandom(data)
         } else {
-          initKMeansParallel(data)
+          try {
+            val exeMemSize = sc.getConf.get("spark.executor.memory")
+            val memDim = exeMemSize.charAt(exeMemSize.length - 1)
+            if ((memDim.equals('g') || memDim.equals('G')) && exeMemSize.length < 3 ||
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: Eh, can you explain what this is? I don't think this even handles all cases.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 2]
```diff
@@ -246,10 +248,23 @@ class KMeans private (
...
+                (memDim.equals('m') || memDim.equals('M')) && exeMemSize.length < 6 ||
+                (memDim.equals('k') || memDim.equals('K')) && exeMemSize.length < 9) {
+              initKMeansParallelFallBack(data)
+            } else {
+              initKMeansParallel(data)
+            }
+          } catch {
+            case _: NoSuchElementException => initKMeansParallel(data)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala]: Why this?

**Comment by srowen** [comment]
I personally think it's brittle to arbitrarily change behavior based on one executor's memory size. It seems like this suggests you can only use the current implementation with >= 10g of memory, which doesn't sound right.

**Comment by HuJiayin** [comment]
Kmeans|| is a better algorithms to find the centers, if user has sufficient memory, the performance is better. But sometime because of Kmeans parameters like K, samples settings and user case related, the algorithms will induce much duplication calculation and cause a large memory consumption in each node. 

I tried the 8G executor memory in each node and 10 slaves in yarn mode. (The dimension is 20, k=5, and then k=50000 samples=1.2billion remains the same.) The task failed with RDD lost after reduced the RDD storage in Kmeans|| but can get the result by using old method. When I enlarged the memory to 20G, the Kmeans|| achieve the same performance with old method.  

The different between kmeans|| and old method is not significant, If user data is small and set a small executor memory. If user set small executor memory in each node and have a large number of vm or nodes, I think the networking cost will raise. It seems not very practicable.

**Comment by HuJiayin** [comment]
Users can manually adjust memory when they meet failure but they may cost some time to find the root cause. There are many ways to implement kmeans and they work well in different cases. I think it maybe better if the switch is transparent to users. The codes can switch between different approaches.

**Comment by HuJiayin** [comment]
The user doesn't need enlarge the memory to run 1.5 kmeans after apply this fix. They still can use 1.2 configuration and have stable run experience in the same time.

**Comment by mengxr** [comment]
@HuJiayin Could you add a description to the PR title? The  JIRA number doesn't describe the content. For the implementation, maybe we can avoid duplicating code. How about not caching the cost RDD if the number of features is small? It should be able to reduce the storage overhead, but it might be slower than before.

**Comment by HuJiayin** [comment]
@mengxr I reduced centers storage and deleted the fallback and duplicate codes. I tested the functionality, performance on my local side and works. Could you give me a code format template for Intellij, I found my code will have many blank spaces after reformat code.

**Comment by mengxr** [comment]
@HuJiayin This basically reverts the behavior back to 1.2. The changes we made in 1.3 is to avoid recomputing distances between old centers and input points during initialization. That is why we need `newCenters`. If you test the current version with a large `k`, you will see the performance difference. Base on our discussion offline, I think there are not much work to do here. The case when the new implementation introduces overhead is when the dataset is really tall and skinny, but we haven't heard negative feedback from practical use cases yet. Do you mind closing this PR for now? Thanks!

**Comment by HuJiayin** [comment]
On the other hand, newcenters will cause a sudden increasing of memory usage, though call clear immediately, but i think it waits for GC to clear. Newcenter will still merge in some duplicate centers. I heard some clusters in a large internet company here set 4-5G yarn executor memory/node and has related usage. You can give a instruction if they meet similar issue.

**Pull request closed without merging**